pymc_extras/statespace/models/structural/components/cycle.py

```diff
@@ -0,0 +1,301 @@
+
+        self.estimate_cycle_length = estimate_cycle_length
+        self.cycle_length = cycle_length
+        self.innovations = innovations
+        self.dampen = dampen
+        self.n_coefs = 1
+
+        k_endog = len(observed_state_names)
+
+        k_states = 2 * k_endog
+        k_posdef = 2 * k_endog
+
+        obs_state_idx = np.zeros(k_states)
+        obs_state_idx[slice(0, k_states, 2)] = 1
+
+        super().__init__(
+            name=name,
+            k_endog=k_endog,
+            k_states=k_states,
+            k_posdef=k_posdef,
+            measurement_error=False,
+            combine_hidden_states=True,
+            obs_state_idxs=obs_state_idx,
+            observed_state_names=observed_state_names,
+        )
+
+    def make_symbolic_graph(self) -> None:
+        Z = np.array([1.0, 0.0]).reshape((1, -1))
+        design_matrix = block_diag(*[Z for _ in range(self.k_endog)])
+        self.ssm["design", :, :] = pt.as_tensor_variable(design_matrix)
+
+        # selection matrix R defines structure of innovations (always identity for cycle components)
+        # when innovations=False, state cov Q=0, hence R @ Q @ R.T = 0
+        R = np.eye(2)  # 2x2 identity for each cycle component
+        selection_matrix = block_diag(*[R for _ in range(self.k_endog)])
+        self.ssm["selection", :, :] = pt.as_tensor_variable(selection_matrix)
+
+        init_state = self.make_and_register_variable(
+            f"{self.name}", shape=(self.k_endog, 2) if self.k_endog > 1 else (self.k_states,)
+        )
+        self.ssm["initial_state", :] = init_state.ravel()
+
+        if self.estimate_cycle_length:
+            lamb = self.make_and_register_variable(f"{self.name}_length", shape=())
+        else:
+            lamb = self.cycle_length
+
+        if self.dampen:
+            rho = self.make_and_register_variable(f"{self.name}_dampening_factor", shape=())
+        else:
+            rho = 1
+
+        T = rho * _frequency_transition_block(lamb, j=1)
+        transition = block_diag(*[T for _ in range(self.k_endog)])
+        self.ssm["transition"] = pt.specify_shape(transition, (self.k_states, self.k_states))
+
+        if self.innovations:
+            if self.k_endog == 1:
+                sigma_cycle = self.make_and_register_variable(f"sigma_{self.name}", shape=())
+                self.ssm["state_cov", :, :] = pt.eye(self.k_posdef) * sigma_cycle**2
+            else:
+                sigma_cycle = self.make_and_register_variable(
+                    f"sigma_{self.name}", shape=(self.k_endog,)
+                )
+                state_cov = block_diag(
+                    *[pt.eye(2) * sigma_cycle[i] ** 2 for i in range(self.k_endog)]
+                )
+                self.ssm["state_cov"] = pt.specify_shape(state_cov, (self.k_states, self.k_states))
+        else:
+            # explicitly set state cov to 0 when no innovations
+            self.ssm["state_cov", :, :] = pt.zeros((self.k_posdef, self.k_posdef))
```

Comment:
Same here: added this clause to make the logic clearer @jessegrabowski . LMK is superfluous